theories/lra.elpi

```diff
@@ -0,0 +1,498 @@
+quote.hyps F [decl _ _ H|Ctx] Type TypeFM TypeF TypeFQ
+    Type'' TypeFM'' TypeF'' TypeFQ'' VM :-
+  quote F H HM' H' HQ' VM, not (H' = {{ X _ _ }}), !,
+  quote.hyps F Ctx Type TypeFM TypeF TypeFQ Type' TypeFM' TypeF' TypeFQ' VM,
+  Type'' = {{ lp:H -> lp:Type' }},
+  TypeFM'' = {{ IMPL lp:HM' None lp:TypeFM' }},
+  TypeF'' = {{ IMPL lp:H' None lp:TypeF' }},
+  TypeFQ'' = {{ lp:HQ' -> lp:TypeFQ' }}.
+quote.hyps F [decl _ _ _|Ctx] Type TypeFM TypeF TypeFQ
+    Type' TypeFM' TypeF' TypeFQ' VM :- !,
+  quote.hyps F Ctx Type TypeFM TypeF TypeFQ Type' TypeFM' TypeF' TypeFQ' VM.
+quote.hyps F [def _ _ _ _|Ctx] Type TypeFM TypeF TypeFQ
+    Type' TypeFM' TypeF' TypeFQ' VM :- !,
+  quote.hyps F Ctx Type TypeFM TypeF TypeFQ Type' TypeFM' TypeF' TypeFQ' VM.
+quote.hyps _ [] Type TypeFM TypeF TypeFQ Type TypeFM TypeF TypeFQ _.
+
+% [exfalso_if_not_prop In Out Bool] changes [In] to [False]
+% when [In] is not a [Prop] (and then set [Bool] to [true])
+pred exfalso_if_not_prop i:term, o:term, o:term.
+exfalso_if_not_prop Type Type {{ false }} :-
+  coq.typecheck Type {{ Prop }} ok.
+exfalso_if_not_prop _ {{ False }} {{ true }}.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Main tactic
+
+% The tactic takes three arguments, two strings [TacF] and [TacR]
+% that are the names of the Ltac1 tactic to call respactively
+% in the [realFieldType] and [realDomainType] case and a third argument [N],
+% passed unchanged as first argument to the previous tactic.
+% The tactics [TacF] or [TacR] will receive seven arguments:
+% - [N] above
+% - [Efalso] a term of type [bool] indicating if the goal was changed to [False]
+% - [Type''] a term of type [Prop], an implication with the goal
+%   and selected hypotheses
+% - [TypeFM'] the reified [Type''] as a [BFormula MFormula Tauto.isProp]
+% - [TypeF'] the reified [Type''] as a [BFormula (Formula Q) Tauto.isProp]
+% - [TypeFQ'] the reified [Type''] as a [Prop] about [Q] (for <= 8.15 compat)
+% - [VM''] a variable map, giving the interpretation to variables in [TypeF']
+%   it is of type [VarMap.t F] where [F] is the carrier for the detected
+%   [realFieldType] or [realDomainType]
+solve (goal Ctx _Trigger Type _Proof [str TacF, str TacR, N] as G) GL :-
+  exfalso_if_not_prop Type Type' Efalso,
+  std.rev Ctx Ctx', !,
+  rfstr Ctx' Type' F, !,
+  quote F Type' TypeFM TypeF TypeFQ VM, !,
+  quote.hyps F Ctx' Type' TypeFM TypeF TypeFQ Type'' TypeFM' TypeF' TypeFQ' VM,
+  !,
+  rf->numDomain F NDF,
+  std.assert-ok!
+    (coq.typecheck TypeFM' {{ BFormula (@Internals.MFormula lp:NDF) isProp }})
+    "The reification produced an ill-typed result, this is a bug.",
+  std.assert-ok!
+    (coq.typecheck TypeF' {{ BFormula (Formula Q) isProp }})
+    "The reification produced an ill-typed result, this is a bug.",
+  rf->sort F FS, list-constant FS VM VM',
+  coq.reduction.vm.norm
+    {{ Internals.vm_of_list lp:VM' }}
+    {{ VarMap.t lp:FS }}
+    VM'',
```

Comment:
The varmap should not be reduced since it may contain arbitrary terms. My understanding is that, since the default conversion compares terms lazily, we do not have to manually unfold `vm_of_list`.